examples/ecs/archetype_invariants.rs

```diff
@@ -0,0 +1,84 @@
+//! Archetype invariants are rules about which combinations of components can coexist.
+//!
+//! An archetype (in the sense that Bevy uses it) is the "unique set of components" that belong to an entity.
+//! These are useful to codify your assumptions about the composition of your entities.
+//! For example, an entity can never have a `Player` component with a `Camera` together,
+//! or a `GlobalTransform` may only be valid in association with a `Transform`.
+//! By constructing `ArchetypeInvariant`s out of `ArchetypeStatement`s,
+//! we can encode this logic into our app.
+//!
+//! Archetype invariants are guaranteed to hold at *all* points during the app's lifecycle;
+//! this is automtically checked on component insertion and removal, including when entities are spawned.
+//! Make sure to test thoroughly when using archetype invariants in production though;
+//! any violations will result in a panic!
+//!
+//! There are many helper methods provided on `ArchetypeInvariant` to help easily construct common invariant patterns,
+//! but we will only be showcasing some of them here.
+//! For a full list, see the docs for `ArchetypeInvariant`.
+use bevy::prelude::*;
+
+fn main() {
+    App::new()
+        // Archetype invariants are constructed in terms of bundles;
+        // use (MyComponent, ) to construct a bundle with a single item.
+        // This invariant ensures that Player and Camera can never be found together.
+        .add_archetype_invariant(ArchetypeInvariant::<(Player,), (Camera,)>::forbids())
+        // This invariant ensures that the presence of a `GlobalTransform` component implies the existence of a `Transform` component.
+        .add_archetype_invariant(ArchetypeInvariant::<(GlobalTransform,), (Transform,)>::requires())
+        // Note that the converse statement isn't automatically true!
```

Comment:
I would like this too, but we couldn't figure out how to express this using a single archetype invariant. Let us know if you have clever logic thoughts, including adding new enum variants.

Reply:
Ah, you're right, I was imagining it as a function that could just `.add_archetype_invariant` twice but it returns one instead. 
I'll think about if this could nicely be expressed in a single invariant. First thought is that if we could negate an invariant, then negating a `forbids` would do it.

Reply:
Agreed that this is unnatural for this simple case. After chatting on discord with @alice-i-cecile, we realized that we should've just been using `atomic` here for single components. After discussion, we came up with a list of things to change in this example to make the intended patterns more clear:

- Change the current requires example to use atomic instead
- Add a Life component to the example, with ArchetypeInvariant::<Player, Life>::requires()
- Add another enum-like set of components to the example, with ArchetypeInvariant::<Player, (Enum1, Enum2, Enum3)>::requires_one()
- Leave the PR as is with regards to supporting symmetric invariants, but leave the door open for better first class support in the future, if it becomes necessary. (With both requirement and forbids working mutually in most cases, I'd like to punt this until after we see how users actually use this feature.)

Reply:
To support symmetrical invariants, `ArchetypeInvariant` could be an enum instead, where a variant is an implication (our current `ArchetypeInvariant`) and the other is an equivalence.

Reply:
> To support symmetrical invariants, ArchetypeInvariant could be an enum instead, where a variant is an implication (our current ArchetypeInvariant) and the other is an equivalence.

This was actually in the original design, but I don't think it's worth adding at this stage.

Regardless, the above list has been fixed as of 9bb0cf42cf768ccca735e1a95a18638daf754223